src/movegen.hpp

```diff
@@ -469,31 +469,41 @@ inline void movegen::whileBitboardAdd(Movelist &movelist, Bitboard mask, T func)
     }
 }
 
-/// @brief all legal moves for a position
+/// @brief generate moves for a position
 /// @tparam c
 /// @tparam mt
 /// @param movelist
 /// @param board
-template <Color::underlying c, movegen::MoveGenType mt>
-inline void movegen::legalmoves(Movelist &movelist, const Board &board, int pieces) {
+template <Color::underlying c, movegen::MoveGenType mt, movegen::Type type>
+inline void movegen::gen_moves(Movelist &movelist, const Board &board, int pieces) {
     /*
      The size of the movelist might not
      be 0! This is done on purpose since it enables
      you to append new move types to any movelist.
     */
-    auto king_sq = board.kingSq(c);
 
     int double_check = 0;
 
-    Bitboard occ_us  = board.us(c);
-    Bitboard occ_opp = board.us(~c);
-    Bitboard occ_all = occ_us | occ_opp;
-
+    Bitboard occ_us    = board.us(c);
+    Bitboard occ_opp   = board.us(~c);
+    Bitboard occ_all   = occ_us | occ_opp;
     Bitboard opp_empty = ~occ_us;
 
-    Bitboard check_mask = checkMask<c>(board, king_sq, double_check);
-    Bitboard pin_hv     = pinMaskRooks<c>(board, king_sq, occ_opp, occ_us);
-    Bitboard pin_d      = pinMaskBishops<c>(board, king_sq, occ_opp, occ_us);
+    auto king_sq = board.kingSq(c);
+
+    Bitboard check_mask;
+    Bitboard pin_hv;
+    Bitboard pin_d;
+
+    if constexpr (type == movegen::Type::LEGAL) {
+        check_mask = checkMask<c>(board, king_sq, double_check);
+        pin_hv     = pinMaskRooks<c>(board, king_sq, occ_opp, occ_us);
+        pin_d      = pinMaskBishops<c>(board, king_sq, occ_opp, occ_us);
+    } else {
+        check_mask = constants::DEFAULT_CHECKMASK;
+        pin_d      = 0;
```

Comment:
I've noticed that in some areas `BitBoard(0)` is used and in other `0` is used. I'm guessing there is no difference but I wonder if it would be more clear to have a dedicated constant. I know it took me a bit to figure out what was going on the first time.

Reply:
ah yeah meh, that 0 is implicitly converted